packages/compass-explain-plan/src/components/explain-toolbar/explain-toolbar.tsx

```diff
@@ -0,0 +1,198 @@
+import type AppRegistry from 'hadron-app-registry';
+
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import {
+  Button,
+  Icon,
+  SegmentedControl,
+  SegmentedControlOption,
+  Overline,
+  Toolbar,
+  css,
+  spacing,
+  withTheme,
+  WarningSummary,
+  ErrorSummary,
+} from '@mongodb-js/compass-components';
+import { useId } from '@react-aria/utils';
+
+const explainToolbarStyles = css({
+  display: 'flex',
+  flexDirection: 'column',
+  alignItems: 'center',
+  gap: spacing[3],
+  padding: spacing[3],
+  boxShadow: '0 2px 2px rgb(0 0 0 / 20%)',
+  zIndex: 10,
+});
+
+const explainQueryBarStyles = css({
+  width: '100%',
+  position: 'relative',
+});
+
+const explainActionsToolbarStyles = css({
+  display: 'flex',
+  alignItems: 'center',
+  width: '100%',
+});
+
+const explainActionsToolbarRightStyles = css({
+  flexGrow: 1,
+  display: 'flex',
+  alignItems: 'center',
+  justifyContent: 'flex-end',
+  gap: spacing[2],
+});
+
+const READ_ONLY_WARNING_MESSAGE =
+  'Explain plans on readonly views are not supported.';
+const OUTDATED_WARNING_MESSAGE = `The explain content is outdated and no longer in sync
+ with the documents view. Press "Explain" again to see the explain plan for
+ the current query.`;
+
+type ExplainToolbarProps = {
+  globalAppRegistry: AppRegistry;
+  localAppRegistry: AppRegistry;
+  darkMode?: boolean;
+  explainErrorMessage?: string;
+  explainResultId: string;
+  onExecuteExplainClicked: (queryBarStoreState: any) => void;
+  showOutdatedWarning: boolean;
+  showReadonlyWarning: boolean;
+  switchToTreeView: () => void;
+  switchToJSONView: () => void;
+  viewType: 'json' | 'tree';
+};
+
+function UnthemedExplainToolbar({
+  globalAppRegistry,
+  localAppRegistry,
+  darkMode,
+  explainResultId,
+  explainErrorMessage,
+  onExecuteExplainClicked,
+  showOutdatedWarning,
+  showReadonlyWarning,
+  switchToTreeView,
+  switchToJSONView,
+  viewType,
+}: ExplainToolbarProps) {
+  const labelId = useId();
+  const controlId = useId();
+
+  const queryBarRef = useRef<{
+    component: React.ComponentType<any>;
+    store: any; // Query bar store is not currently typed.
+    actions: any; // Query bar actions are not typed.
+  } | null>(null);
+
+  const [queryBarLoaded, setQueryBarLoaded] = useState(false);
+
+  useEffect(() => {
+    const queryBarRole = localAppRegistry.getRole('Query.QueryBar')![0];
+
+    queryBarRef.current = {
+      component: queryBarRole.component,
+      store: localAppRegistry.getStore(queryBarRole.storeName!),
+      actions: localAppRegistry.getAction(queryBarRole.actionName!),
+    };
+
+    setQueryBarLoaded(true);
+  }, [localAppRegistry]);
+
+  const toggleView = useCallback(() => {
+    if (viewType === 'json') {
+      switchToTreeView();
+    } else {
+      switchToJSONView();
+    }
+  }, [viewType, switchToJSONView, switchToTreeView]);
```

Comment:
I would really suggest you do this based on the value passed to the callback to make sure that UI is 100% in sync with the state